File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;

class Solution {
    public long solution(int n, int[] times) {
        long answer = 0;
        
        Arrays.sort(times);
        
        long start = 0;
        long end = (long)times[times.length - 1] * n;
        
        while (start <= end) {
            long mid = (start + end) / 2;
            
            long cnt = 0;
            for (int t : times) {
                cnt += mid / t;
            }
            
            if (cnt >= n) {
                end = mid - 1;
            }else{
                start = mid + 1;
            }
        }
        
        
        return start;
    }
}

File: 프로그래머스/3/43238. 입국심사/입국심사.cpp
#include <bits/stdc++.h>

using namespace std;
typedef long long ll;

ll l, r, mid,INF = 1e18, ret = INF;

bool check(int n, vector<int> times, ll target) {
    ll cnt = 0;

    for (ll t : times) {
        cnt += target / t;
    }

    return cnt >= n;
}

ll solution(int n, vector<int> times) {

    l = 1;
    for (ll t : times) {
        r = max(r, t);
    }
    r *= n;

    while (l <= r) {
        mid = (l + r) / 2;

        if (check(n, times, mid)) {
            ret = min(ret, mid);
            r = mid - 1;
        } else {
            l = mid + 1;
        }
    }



    return ret;
}